CrossrefReferenceLinkingPlugin.php

```diff
@@ -0,0 +1,474 @@
+	 *  @option Citation
+	 *  @option Smarty
+	 *  @option string Rendered smarty template
+	 * ]
+	 */
+	public function displayReferenceDOI(string $hookName, array $params): bool
+	{
+		$citation =& $params[0]['citation'];
+		$smarty =& $params[1];
+		$output =& $params[2];
+
+		if ($citation->getData($this->getCitationDoiSettingName())) {
+			$crossrefFullUrl = 'https://doi.org/' . $citation->getData($this->getCitationDoiSettingName());
+			$smarty->assign('crossrefFullUrl', $crossrefFullUrl);
+			$output .= $smarty->fetch($this->getTemplateResource('displayDOI.tpl'));
+		}
+		return false;
+	}
+
+	/**
+	 * Get citations diagnostic ID setting name.
+	 */
+	public function getCitationsDiagnosticIdSettingName(): string
+	{
+		return 'crossref::citationsDiagnosticId';
+	}
+
+	/**
+	 * Get citation crossref DOI setting name.
+	 */
+	public function getCitationDoiSettingName(): string
+	{
+		return 'crossref::doi';
+	}
+
+	/**
+	 * Get setting name, that defines if the scheduled task for the automatic check
+	 * of the found Crossref citations DOIs should be run, if set up so in the plugin settings.
+	 */
+	public function getAutoCheckSettingName(): string
+	{
+		return 'crossref::checkCitationsDOIs';
+	}
+
+	/**
+	 * Retrieve all submissions that should be automatically checked for the found Crossref citations DOIs.
+	 */
+	public function getSubmissionsToCheck(Context $context): array
+	{
+		// Retrieve all published articles with their DOIs depositted together with the references.
+		// i.e. with the citations diagnostic ID setting
+		$submissionIds = Repo::submission()->getIdsBySetting($this->getAutoCheckSettingName(), true, $context->getId())->toArray();
+		$submissions = array_map(function($submissionId) {
+			return Repo::submission()->get($submissionId);
+		}, $submissionIds);
+
+		return array_filter($submissions, function($submission) {
+			return $submission->getData('status') === Submission::STATUS_PUBLISHED;
+		});
+	}
+
+	/**
+	 * Use Crossref API to get the references DOIs for the the given article DOI.
+	 */
+	protected function _getResolvedRefs(string $doi, int $contextId): ?array
+	{
+		$matchedReferences = null;
+
+		PluginRegistry::loadCategory('generic');
+		$crossrefPlugin = PluginRegistry::getPlugin('generic', 'crossrefplugin');
+		$username = $crossrefPlugin->getSetting($contextId, 'username');
+		$password = $crossrefPlugin->getSetting($contextId, 'password');
+
+		// Use a different endpoint for testing and production.
+		$isTestMode = $crossrefPlugin->getSetting($contextId, 'testMode') == 1;
+		$endpoint = ($isTestMode ? self::CROSSREF_API_REFS_URL_DEV : self::CROSSREF_API_REFS_URL);
+
+		$url = $endpoint.$doi.'&usr='.$username.'&pwd='.$password;
+
+		$httpClient = Application::get()->getHttpClient();
+		try {
+			$response = $httpClient->request('POST', $url);
+		} catch (\GuzzleHttp\Exception\RequestException $e) {
+			return null;
+		}
+
+		if ($response && $response->getStatusCode() == 200)  {
```

Comment:
I think the `$response` should be not null here, otherwise, you could use `$response->?getStatusCode()`.